Use ZStandard Compression for PARK and PARKREP Files

LRFLEW opened the pull request:
When it comes to comparing compression algorithms, there are largely two important metrics: compression (and decompression) speed, and compression ratio. Compared to Zlib, which is the current compression library used by OpenRCT2, most algorithms only improve on it in one of these factors, such as LZ4 being faster with a worse compression ratio, or LZMA having a better compression ratio at a slower speed. However, the relatively new compression algorithm [ZStandard](https://github.com/facebook/zstd) (Zstd, not to be confused with Zlib) developed by Facebook is generally both faster than Zlib and produces smaller results at the same time. Both Zlib and Zstd have multiple compression levels to tune the priority between speed and compression ratio, but Zstd generally has a compression level that is strictly better than any given Zlib level.

![Graph comparing Zstd and Zlib compression levels, showing Zstd compression being both faster and smaller than Zstd compression.](https://github.com/facebook/zstd/blob/dev/doc/images/CSpeed2.png?raw=true)
(Provided by the Zstd project's [Github page](https://github.com/facebook/zstd?tab=readme-ov-file#benchmarks))

I had been thinking about experimenting with adding Zstd to OpenRCT2 for a while, knowing that both compression speed and ratio would have a significant impact on networking, as improving both of these would decrease the time it takes for the client to receive the park data, and therefore how much catching up it has to do. While investigating it, I saw that the PARK format has a field for the type of compression used, which makes it simple to just add a new value there to indicate Zstd compression. Similarly, I noticed that Zlib was being used for the replay files, so I decided to switch that one over as well.

One question then is what compression level to use with Zstd. The default level is 3, but the standard levels range from 1 to 19. (The levels do go lower with the negative "fast" levels, as well as the higher "ultra" levels 20-22. However, the negative levels aren't very useful for this, and the ultra levels are not recommended by Facebook due to their increased memory usage.) Since the best level to use for people may vary by machine, I decided to add config options to config.ini for each of the cases that a save file is generated (so separate options for auto saves, manual saves, saves for network transfers, and replay file saves). I didn't add options to the UI to adjust these, as these would only be power-user options, and the majority of users shouldn't ever adjust these. The question, then, is what default levels should these configs use.

To test how the different compression levels perform, it helps to have a representative sample of the data being compressed. With that in mind, I downloaded a ton of .park save files off of RCTgo, collecting 477 files that should be fairly representative. I then extracted and decompressed the chunks section of these files to use for benchmarking. I then ran the benchmarking tool built into the `zstd` command-line tool on all the files with all 19 levels, averaged the result across the files, and analyzed the result. This is the graph I got from the results of this testing (x axis is in MB/s, and y axis is compressed size / uncompressed size):

![Graph showing how Zstd performs at different compression levels with OpenRCT2 park file data](https://github.com/user-attachments/assets/10a9464e-954e-4850-ad9f-7b3d7f4f0b98)

With this information, I chose some different default compression levels for the different save cases. I then tested these specific levels against Zlib using another tool that supported both compression types (lzbench) to get the relative performance of making the switch from Zlib to Zstd.

* For Network and Automatic Saves, a good balance of speed and size is needed. For networking, smaller sizes means less time spent transferring data, but only if you're not spending more time compressing than you save in transferring time. As for auto-saves, decreasing the file size is useful for reducing the storage space used and reducing the time spent writing, but the longer spent compressing means the game can stutter for longer during the save. (Even if the auto-save compression and write is performed on a separate thread, it can still impact performance due to things like scheduling and reduced CPU clock rates.) The default compression level of 3 is a good choice for a balance between speed and compression ratio. However, the testing showed that level 4 is **very** similar performance-wise to level 3 for this test data, so I opted to go with level 4 instead for the slight size improvement. Compared to Zlib default level compression (which this replaces), Zstd level 4 compresses 12.8 times faster, and produces 12.5% smaller results.
* For manual saves, getting a smaller file is more important than speed. Since manual saves are triggered directly by the user, it's reasonable for the game to take longer compressing the data. Additionally, manual saves are generally intended to be kept for longer than auto-saves, so the resulting file size is more important for them. However, it's not worth it to dramatically slow down the compression speed for minimal size improvements, so a reasonable compromise is needed here. For this case, I chose a default compression level of 7, since the compression ratio levels off a lot after that level. Compared to Zlib default level compression, Zstd level 7 compresses 4.5 times faster, and produces 20.5% smaller results.
* For replay files, this gets a bit more complicated. I don't have access to a good representative sample of these files. However, with the change made in #24728, the largest portion of these files will be the chunks from the park data, so the sample data used here should hopefully be reasonably representative. Since this currently uses max-level Zlib compression, I assume this is a case where higher compression ratios is more important than speed. An obvious choice would be Zstd level 19 compression (or higher if we're ok with "ultra" levels), but based on the results of the testing, I instead went with level 18. This is because the results showed a bit of a falling off in the ratio going up from 18, and level 18 is already a significant upgrade. Compared to Zlib max level compression, Zstd level 18 compresses 1.7 times faster, and produces 28.0% smaller results.

The decompression speeds of both Zstd and Zlib don't change significantly with compression level, but in all three tested cases, Zstd decompresses about 3.5 to 3.9 times faster than Zlib, which is also useful.

There is certainly room here for more testing or fine tuning these compression levels based on feedback, so feel free to let me know if you want me to adjust these values or perform more tests. I can also remove the INI file options for these levels if you'd rather these levels simply be hard-coded into the game.

Since using a new compression algorithm breaks compatibility with older builds, I bumped the version numbers for the PARK files, the PARKREP files, and the networking version. Notably, the use of a variable for the compression type in the PARK files means I *could* have avoided bumping that version number. However, prior to #24701, OpenRCT2 didn't properly handle invalid compression types in the PARK header, so bumping the version number prevents builds from before that change from misinterpreting the PARK file. If a new compression type is added in the future, or `kParkFileMinVersion` gets bumped before this is merged, then the PARK file version does not need to be bumped again.

I did keep support for loading Gzip-compressed PARK and PARKREP files using Zlib for backwards compatibility. I don't see any reason that we would consider dropping this in the future, as we will still depend on Zlib to support loading things like PNG and ZIP files. This also still continues to use Gzip compression for crash logs. That *can* be switched over to using Zstd compression (with the file extension .zst), but before doing so, we should check if it's supported not only with things like backtrace.io, but also with common decompressing tools, especially those people might be using on Windows and macOS. For now, I'm keeping the crash log compression as Gzip compression.

Lastly, I went ahead and updated the `convert` command-line command to allow for updating PARK files with the new compression to make it easier to test this change. I also added a new argument `--compress-level` (or `-l`) that lets you specify a compression level to use. This could be useful to, for example, re-compress park files with a higher compression level before releasing a new title sequence.

ZehMatt commented on src/openrct2/core/Compression.cpp:
```diff
@@ -138,4 +139,137 @@ namespace OpenRCT2::Compression
+     * Modified copy of ZSTD_COMPRESSBOUND / ZSTD_compressBound() from zstd 1.5.7, with the argument
+     * type changed from size_t (which may be only 32 bits) to uint64_t, and removes the error handling.
+     */
+    static uint64_t zstdCompressBound(uint64_t length)
```
Why do we need this to be 64 bit? Nothing should ever get this far, if it does then we are doing something wrong.

ZehMatt replied:
Just not a fan of duplicating code, the existing one is perfectly usable until it goes beyond 4 GiB which shouldn't happen.

ZehMatt replied:
After a chat on Discord, I think its okay, its not absolutely necessary but helps avoid pitfalls.

ZehMatt commented on test/tests/S6ImportExportTests.cpp:
```diff
@@ -120,7 +120,7 @@ static bool ExportSave(MemoryStream& stream, std::unique_ptr<IContext>& context)
 
     auto& gameState = getGameState();
-    exporter->Export(gameState, stream);
+    exporter->Export(gameState, stream, 7);
```
Magic number, should have a constant.

LRFLEW replied:
I added a constant to the test file itself. Not sure if that's the best solution, so I'm gonna keep this marked as unresolved for the time being.

LRFLEW commented:
On reconsideration, I ended up agreeing with @ZehMatt's comment, and removed the config values for setting the compression levels. While I figured it could be useful for power-users, I realized the same could be instead handled by having those users either recompile with patched values or using the convert command, depending on the use-case. Making this change also simplifies the code for things like the convert command, since it no longer needs to load in the config file to work properly.

With the PRs for the docker images and Android dependencies merged, I switched the current commit to using the final upstream references. Since the changes for the Windows and macOS dependencies (https://github.com/OpenRCT2/Dependencies/pull/44) hasn't been merged and made a release, I still have the temporary commit to get those to build. Other than that, the only blocker right now is the AppImage build not working on the 20-noble image. I'll see if I can get that working and make a separate PR for it.

janisozaur commented:
Depends on #24887 and #24889

janisozaur commented:
With the amount of dependency updates we had to do to add zstd, I now realise we could have added SDL3 at the same time.

LRFLEW commented:
> Depends on #24887 and #24889

I think I may have duplicated that effort in this PR already, lol. Either way, that should work fine.

> With the amount of dependency updates we had to do to add zstd, I now realise we could have added SDL3 at the same time.

I figured it would be safer to only add SDL3 once we're ready to remove SDL2, since I don't know if there might be conflicts with having both "installed" at the same time.

LRFLEW commented:
Also, since this affects save forwards compatibility (i.e. loading new saves in old builds), should I add a changelog entry for this change in this PR?

janisozaur commented:
>  I downloaded a ton of .park save files off of RCTgo, collecting 477 files that should be fairly representative

Can you share those parks?

janisozaur commented on src/openrct2/core/Compression.cpp:
```diff
@@ -19,20 +19,21 @@
-    constexpr int kZlibWindowBits[] = { -15, 15, 15 + 16 };
+    static constexpr size_t kZlibChunkSize = 128 * 1024;
+    static constexpr size_t kZlibMaxChunkSize = static_cast<size_t>(std::numeric_limits<uInt>::max());
+    static constexpr int kZlibWindowBits[] = { -15, 15, 15 + 16 };
```
What's the static here for?

LRFLEW replied:
This was something I meant to do back when I first added these constants, but forgot. I figured I could do it here like with adding the `uLL` suffixes that @ZehMatt suggested. Short version: since these values are only declared in the source file, they should have internal linkage, meaning declared in an anonymous namespace or using the static keyword. It shouldn't affect how it works, but it's a change still worth making.

janisozaur commented on src/openrct2/core/Compression.cpp:
```diff
@@ -138,4 +139,137 @@ namespace OpenRCT2::Compression
+     * Modified copy of ZSTD_COMPRESSBOUND / ZSTD_compressBound() from zstd 1.5.7, with the argument
+     * type changed from size_t (which may be only 32 bits) to uint64_t, and removes the error handling.
+     */
+    static uint64_t zstdCompressBound(uint64_t length)
```
zstdCompressBound64?

LRFLEW replied:
I could name it something like that. The current name isn't going to have collisions with the zstd library, since all its function names start with `ZSTD_`, so changing that would only be a stylistic change. I don't see much reason to make the change, though. If I change the name of this function, we should also rename [`zlibCompressBound`](https://github.com/OpenRCT2/OpenRCT2/blob/b9c3306c3d7c0644e572873fb2330407d2c4a043/src/openrct2/core/Compression.cpp#L33-L36).

janisozaur commented on src/openrct2/core/Compression.cpp:
```diff
@@ -138,4 +139,137 @@ namespace OpenRCT2::Compression
+        size_t ret;
+        StreamReadBuffer sourceBuf(source, sourceLength, ZSTD_CStreamInSize());
+        StreamWriteBuffer destBuf(dest, zstdCompressBound(sourceLength), ZSTD_CStreamOutSize());
+        unsigned metaFlags = static_cast<unsigned>(metadata);
```
uint32_t?

LRFLEW replied:
I could switch this type, but I had used `unsigned` for a reason. Specifically, `metadata`'s type is the enum type `ZstdMetadata`, which uses the default underlying type of `int`. Since I am performing bitwise operations on the result, I prefer having an unsigned result (since it's less likely to have undefined behavior). As such, the type to cast to should be `std::make_unsigned_t<std::underlying_type_t<ZstdMetadata>>`, which is `unsigned`. If I *do* change this to `uint32_t` (for stylistic reasons I assume), I should also change the underlying type of `ZstdMetadata` to `[u]int32_t` to match.

janisozaur commented:
List of PRs required to make this happen:

https://github.com/OpenRCT2/openrct2-docker/pull/27
https://github.com/OpenRCT2/Dependencies/pull/44
https://github.com/OpenRCT2/openrct2-dependencies-android/pull/17
https://github.com/OpenRCT2/openrct2-dependencies-android/pull/19
https://github.com/OpenRCT2/openrct2-dependencies-android/pull/20
https://github.com/OpenRCT2/openrct2-docker-build/pull/21
https://github.com/OpenRCT2/openrct2-docker-build/pull/22
https://github.com/OpenRCT2/OpenRCT2/pull/24887
https://github.com/OpenRCT2/OpenRCT2/pull/24889
https://github.com/OpenRCT2/OpenRCT2/pull/24868
https://github.com/OpenRCT2/OpenRCT2/pull/24862
